zaza/utilities/run_report.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2019 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for logging events for a summary report.
+
+This module contains a number of functions for logging events so that they
+can be summarised, with timings, at the end of the run.
+"""
+
+import copy
+import logging
+import time
+import math
+import yaml
+
+
+start = 'Start'
+finish = 'Finish'
+metadata = 'Metadata'
+events = 'Events'
+pct_of_runtime = 'PCT Of Run Time'
+elapsed_time = 'Elapsed Time'
+
+run_data = None
+
+event_states = [start, finish]
+
+
+def _init_run_data(reset=False):
+    global run_data
+    if not run_data or reset:
+        run_data = {
+            events: {},
+            metadata: {}}
+
+
+def register_event(event_name, event_state, timestamp=None):
+    """Register that event_name is at event_state.
+
+    :param event_name: Name of event
+    :type event_name: str
+    :param event_state: Name of event state (should be one of event_states).
+    :type event_state: str
+    :param timestamp: Seconds since epoch when event_state of event_name was
+                      reached.
+    :type timestamp: float
+    """
+    global run_data
+    _init_run_data()
+    timestamp = timestamp or time.time()
+    if event_name in run_data[events]:
+        run_data[events][event_name][event_state] = timestamp
+    else:
+        run_data[events][event_name] = {event_state: timestamp}
+
+
+def register_metadata(cloud_name=None, model_name=None, target_bundle=None):
```

Comment:
Could do this more generically as `**kwargs` and then just configure metadata based off of that, rather than having a restricted set of options

Reply:
My github foo seems to be weak and I've lost my comment:  I went back and forth with making thisjust use kwargs. In then end I figured it might be useful to enforce some consistent key naming. We can always loosen this up later.